### ISSUE TITLE
Progress - NullReferenceException

### ISSUE THREAD

**Issue by jcox86** [aritchie/userdialogs#4]
I'm trying to show a progress indicator while building some tables in the local cache.
```c#
var config = new ProgressDialogConfig
{
    Title = "Refreshing Local Cache",
    AutoShow = true
};
using (var dialog = UserDialogs.Instance.Progress(config))
{
    await _cacheBuilder.BuildCacheAsync(message.ForceRebuild);
    dialog.Hide();
}
```
However when this runs I get a NullReferenceException on this line
```c#
using (var dialog = UserDialogs.Instance.Progress(config))
```

I am unable to get the progress dialog to show at all, as it never hits the next line.

This is the stack trace I am getting pointing to the BigTed.ProgressHUD
```
{System.NullReferenceException: Object reference not set to an instance of an object
  at BigTed.ProgressHUD.PositionHUD (Foundation.NSNotification notification) [0x00000] in <filename unknown>:0 
  at BigTed.ProgressHUD.ShowProgressWorker (Single progress, System.String status, MaskType maskType, Boolean textOnly, ToastPosition toastPosition, System.String cancelCaption, System.Action cancelCallback, Double timeoutMs, Boolean showContinuousProgress, UIKit.UIImage displayContinuousImage) [0x00000] in <filename unknown>:0 
  at BigTed.ProgressHUD+<Show>c__AnonStorey0.<>m__0 () [0x00000] in <filename unknown>:0 
  at Foundation.NSActionDispatcher.Apply () [0x00000] in /Developer/MonoTouch/Source/maccore/src/Foundation/NSAction.cs:56 
  at (wrapper managed-to-native) ObjCRuntime.Messaging:void_objc_msgSend_IntPtr_IntPtr_bool (intptr,intptr,intptr,intptr,bool)
  at Foundation.NSObject.InvokeOnMainThread (System.Action action) [0x00007] in /Developer/MonoTouch/Source/maccore/src/Foundation/NSObject2.cs:575 
  at BigTed.ProgressHUD.Show (System.String status, Single progress, MaskType maskType, Double timeoutMs) [0x00000] in <filename unknown>:0 
  at BigTed.BTProgressHUD.Show (System.String status, Single progress, MaskType maskType) [0x00000] in <filename unknown>:0 
  at Acr.UserDialogs.ProgressDialog.Refresh () [0x00000] in <filename unknown>:0 
  at Acr.UserDialogs.ProgressDialog.Show () [0x00000] in <filename unknown>:0 
  at Acr.UserDialogs.AbstractUserDialogs.Progress (Acr.UserDialogs.ProgressDialogConfig config) [0x00000] in <filename unknown>:0 
  at Projects.Features.BuildCache.BuildAccountCacheCommandHandler+<HandleAsync>d__1.MoveNext () [0x000b5] in c:\Projects\xamarin\Features\BuildCache\BuildAccountCacheCommandHandler.cs:46 
--- End of stack trace from previous location where exception was thrown ---
```
All other UserDialogs elements show up just fine, so what is different about what I am doing with Progress?

**Comment by aritchie** [comment]
Without more details, I'm going to assume you are using xamarin forms.  If you are, do not initialize a loader in the constructor.  You have to do it with the page.OnAppearing.

**Comment by supreettare** [comment]
Did you initiate the UserDialog in AppDelegate? Probably that will fix the issue. 
In AppDelegate.cs in IOS write this 
UserDialogs.init(); 

This should fix the issue.

**Comment by jcox86** [comment]
Sorry, I am using Xamarin Forms 1.3.4.6331-pre4 and Acr.UserDialogs 3.0.3

I am using the MessagingCenter to publish and subscribe to events.  In this case I am subscribing to a log in command, which in the handler, I am trying to perform the above of displaying a progress dialog while I build the cache.

@supreettare I am already doing the Init() in the AppDelegate.  All my other UserDialogs work, ActionSheet, Toast, etc.

**Comment by aritchie** [comment]
I'm still guessing here, but if you are publishing an event to a backgrounded view and you are starting your load process before it is visible again, you are going to have issues like this because the hud can't figure out where to draw itself.

In your subscription, you should flag your view to refresh itself when it becomes visible again.

**Comment by jcox86** [comment]
I found the issue.  I was loading a splash page while calling this async method and thought the splash was available for the progress dialog to position itself, but the UI thread had not yet returned and thus had nothing to display on, like you said.
Thanks for your help.

**Comment by whitewolf33** [comment]
Hi Guys, Stuck with this issue for the past 2 days. No clue what is causing this. Very similar to how jcox86 had mentioned but still no luck. Disabled all progress, tried making the async call to wait to come back yet in vain. Just trying to post it here if in case you know the answer to it or how to solve this. 

App: XamForms, Acr.XamForms.UserDialogs.iOS 3.0.3

App Start -> Splash Page -> OnAppearing -> Call a AsyncMethod -> Even before the Async method finishes, App Crashes.

Oct 13 11:29:11 My-iPhone MyAppiOS[2437] <Warning>: Unhandled managed exception:
	Object reference not set to an instance of an object (System.NullReferenceException)
	  at BigTed.ProgressHUD.<DismissWorker>m__5 () [0x000fc] in <filename unknown>:0 
	  at Foundation.NSActionDispatcher.Apply () [0x00000] in /Users/builder/data/lanes/2077/d8e9592a/source/maccore/src/Foundation/NSAction.cs:57 
	  at (wrapper managed-to-native) ObjCRuntime.Messaging:void_objc_msgSend_IntPtr_IntPtr_bool (intptr,intptr,intptr,intptr,bool)
	  at Foundation.NSObject.InvokeOnMainThread (System.Action action) [0x00007] in /Users/builder/data/lanes/2077/d8e9592a/source/maccore/src/Foundation/NSObject2.cs:564 
	  at BigTed.ProgressHUD.<DismissWorker>m__4 () [0x0003b] in <filename unknown>:0 
	  at UIKit.UIView+<Animate>c__AnonStorey2.<>m__0 (Boolean x) [0x0000b] in /Users/builder/data/lanes/2077/d8e9592a/source/maccore/src/UIKit/UIView.cs:134 
	  at ObjCRuntime.Trampolines+SDUICompletionHandler.Invoke (IntPtr block, Boolean finished) [0x00019] in /Users/builder/data/lanes/2077/d8e9592a/source/maccore/src/build/ios/native/ObjCRuntime/Trampolines.g.cs:15970 
	  at (wrapper native-to-managed) ObjCRuntime.Trampolines/SDUICompletionHandler:Invoke (intptr,int)
	  at (wrapper managed-to-native) UIKit.UIApplication:UIApplicationMain (int,string[],intptr,intptr)
	  at UIKit.UIApplication.Main (System.String[] args, IntPtr principal, IntPtr delegate) [0x00005] in /Users/builder/data/lanes/2077/d8e9592a/source/maccore/src/UIKit/UIApplication.cs:74 
	  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0001c] in /Users/builder/data/lanes/2077/d8e9592a/source/maccore/src/UIKit/UIApplication.cs:57 
	  at MyApp.iOS.Application.Main (System.String[] args) [0x00033] in d:\Working\Source\MyApp\MyApp.iOS\Main.cs:25

**Comment by aritchie** [comment]
@whitewolf33 It looks like you are calling for the dialogs before the UI thread is available.

**Comment by whitewolf33** [comment]
This just happens on iOS 9.0.X not in other versions including 9.1 beta. 

I tried to make the UI thread wait , forced the dialog to be run on the main thread but no luck at all. Will let you know if i succeed but most likely would be going without the dialogs. 

Love them though. Appreciate your efforts and the amazing dialogs

Regards

Vinod Srinivasan

**Comment by aritchie** [comment]
The dialogs will dispatch to the UI thread internally.  Can you supply a test case with the issue?

**Comment by whitewolf33** [comment]
Hi aritchie, I will try and get a sample App done and will let you know. Thanks for reopening this for me.Will also give me a chance to rewrite and see if it is a genuine issue. Just busy with work. Please bear with me until early next week. - Vinod

**Comment by whitewolf33** [comment]
Hi Allan;

            Apologies for not getting back to you on this sooner. I have identified that this issue is very specific to iOS 9.0.x and with the latest 9.1 the crash does not occur any more.

Another App (not the one I identified this issue in) had reported the same crash from one of the users running iOS 9.0.2 and before I could try and get to the bottom of it, I tried the latest 9.1 released and this issue does not occur. I guess it is not worth spending any more time to investigate on this, unless you want to, as this is specific only to iOS 9.0.x and not in versions prior or above it.

Thank you for reopening this issue and showing your support. Much appreciated. I have no problem if you go ahead and close this issue.
